File: src/CCVTAC.Console/IoUtilities/Directories.cs
using System.IO;
using System.Text;
using CCVTAC.Console.PostProcessing;

namespace CCVTAC.Console.IoUtilities;

internal static class Directories
{
    private const string AllFilesSearchPattern = "*";
    private static readonly EnumerationOptions EnumerationOptions = new();

    internal static int AudioFileCount(string directory)
    {
        return new DirectoryInfo(directory)
            .EnumerateFiles()
            .Count(f => PostProcessor.AudioExtensions.CaseInsensitiveContains(f.Extension));
    }

    internal static Result WarnIfAnyFiles(string directory, int showMax)
    {
        var fileNames = GetDirectoryFileNames(directory);
        var fileCount = fileNames.Length;

        if (fileNames.IsEmpty)
        {
            return Result.Ok();
        }

        var fileLabel = fileCount == 1 ? "file" : "files";
        var report = new StringBuilder($"Unexpectedly found {fileCount} {fileLabel} in working directory \"{directory}\":{Environment.NewLine}");

        foreach (string fileName in fileNames.Take(showMax))
        {
            report.AppendLine($"• {fileName}");
        }

        if (fileCount > showMax)
        {
            report.AppendLine($"... plus {fileCount - showMax} more.");
        }

        return Result.Fail(report.ToString());
    }

    internal static Result<int> DeleteAllFiles(string workingDirectory, int showMaxErrors)
    {
        var fileNames = GetDirectoryFileNames(workingDirectory);

        int successCount = 0;
        var errors = new List<string>();

        foreach (var fileName in fileNames)
        {
            try
            {
                File.Delete(fileName);
                successCount++;
            }
            catch (Exception ex)
            {
                errors.Add(ex.Message);
            }
        }

        if (errors.Count == 0)
        {
            return Result.Ok(successCount);
        }

        var output = new StringBuilder($"While {successCount} files were deleted successfully, some files could not be deleted:");
        foreach (string error in errors.Take(showMaxErrors))
        {
            output.AppendLine($"• {error}");
        }

        if (errors.Count > showMaxErrors)
        {
            output.AppendLine($"... plus {errors.Count - showMaxErrors} more.");
        }

        return Result.Fail(output.ToString());
    }

    internal static Result<int> AskToDeleteAllFiles(string workingDirectory, Printer printer)
    {
        bool doDelete = printer.AskToBool("Delete all temporary files?", "Yes", "No");

        return doDelete
            ? DeleteAllFiles(workingDirectory, 10)
            : Result.Fail("Will not delete the files.");
    }

    /// <summary>
    /// Returns the filenames in a given directory, optionally ignoring specific filenames.
    /// </summary>
    /// <param name="directoryName"></param>
    /// <param name="customIgnoreFiles">An optional list of files to be excluded.</param>
    private static ImmutableArray<string> GetDirectoryFileNames(
        string directoryName,
        IEnumerable<string>? customIgnoreFiles = null)
    {
        var ignoreFiles = customIgnoreFiles?.Distinct() ?? [];

        return
            [
                ..Directory
                    .GetFiles(directoryName, AllFilesSearchPattern, EnumerationOptions)
                    .Where(filePath => ignoreFiles.None(filePath.EndsWith))
            ];
    }
}

File: src/CCVTAC.FSharp/IoUtilities/Directories.fs
namespace CCVTAC.Console.IoUtilities

open CCVTAC.Console
open System.IO

module Directories =

    [<Literal>]
    let private allFilesSearchPattern = "*"

    /// Counts the number of audio files in a directory.
    let audioFileCount (directory: string) (includedExtensions: string list) =
        DirectoryInfo(directory).EnumerateFiles()
        |> Seq.filter (fun f -> List.caseInsensitiveContains f.Extension includedExtensions)
        |> Seq.length

    /// Returns the filenames in a given directory, optionally ignoring specific filenames.
    let private getDirectoryFileNames
        (directoryName: string)
        (customIgnoreFiles: string seq option)
        : Result<string array, string> =

        let ignoreFiles =
            customIgnoreFiles
            |> Option.defaultValue Seq.empty
            |> Seq.distinct
            |> Seq.toArray

        ofTry (fun _ ->
            Directory.GetFiles(directoryName, allFilesSearchPattern, EnumerationOptions())
            |> Array.filter (fun filePath -> not (ignoreFiles |> Array.exists filePath.EndsWith)))

    /// Empties a specified directory and reports the count of deleted files.
    let deleteAllFiles showMaxErrors workingDirectory : Result<uint,string> =
        let delete fileNames =
            Array.fold
                (fun (successCount: uint, errors: string list) fileName ->
                    try
                        File.Delete fileName
                        (successCount + 1u, errors)
                    with exn -> (successCount, errors @ [exn.Message]))
                (0u, [])
                fileNames

        match getDirectoryFileNames workingDirectory None with
        | Error errMsg -> Error errMsg
        | Ok fileNames ->
            match delete fileNames with
            | successCount, [] ->
                Ok successCount
            | successCount, errors ->
                SB($"{String.fileLabel None successCount} deleted successfully, but some files could not be deleted:{String.newLine}")
                    .AppendLine(fileNames
                                |> Array.truncate showMaxErrors
                                |> Array.map (sprintf "• %s")
                                |> String.concat String.newLine)
                |> fun sb ->
                    if errors.Length > showMaxErrors
                    then sb.AppendLine $"... plus {errors.Length - showMaxErrors} more."
                    else sb
                |> _.ToString()
                |> Error

    /// Ask the user to confirm the deletion of files in the specified directory.
    let askToDeleteAllFiles dirName (printer: Printer) =
        if printer.AskToBool("Delete all temporary files?", "Yes", "No")
        then deleteAllFiles 10 dirName
        else Error "Will not delete the files."

    /// Warn the user if there are any files in the specified directory.
    let warnIfAnyFiles showMax dirName =
        match getDirectoryFileNames dirName None with
        | Error errMsg -> Error errMsg
        | Ok fileNames ->
            if Array.isEmpty fileNames then
                Ok ()
            else
                SB($"Unexpectedly found {String.fileLabel None fileNames.Length} in working directory \"{dirName}\":{String.newLine}")
                    .AppendLine
                        (fileNames
                         |> Array.truncate showMax
                         |> Array.map (sprintf "• %s")
                         |> String.concat String.newLine)
                |> fun sb ->
                    if fileNames.Length > showMax
                    then sb.AppendLine $"... plus {fileNames.Length - showMax} more."
                    else sb
                |> _.AppendLine("This sometimes occurs due to the same video appearing twice in playlists.")
                |> _.ToString()
                |> Error

    /// Ensures the specified directory exists, including creation of it if necessary.
    let ensureDirectoryExists dirName : Result<DirectoryInfo, string> =
        try
            dirName
            |> Path.GetFullPath
            |> Directory.CreateDirectory
            |> Ok
        with exn ->
            Error $"Error accessing or creating directory \"%s{dirName}\": %s{exn.Message}"

